### PR TITLE
Postgres

### PR DESCRIPTION
**Pull request by vikramsg**
Add FastAPI app that does CRUD operations on a PostGres tables

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,9 @@ old
 *.swp
 *.mod
 *.out
+*.pyc
+.env
+.venv
+local_pgdata
+pgadmin-data
+postgres/src/__pycache__/__init__.cpython-38.pyc
```
Comment (pt20) [inline .gitignore:12]: `.vscode`

File: postgres/.vscode/settings.json
Comment (pt20) [inline postgres/.vscode/settings.json:1]: remove this file by putting it in gitignore. See the comment on the gitignore.

File: postgres/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+RUN --mount=type=cache,target=/home/.cache/pypoetry/cache \
+    --mount=type=cache,target=/home/.cache/pypoetry/artifacts \ 
+    poetry install --only main
+ENV PYTHONPATH /app
+
+# Expose the port on which the application will run (default: 8000)
+EXPOSE 8000
+
+# Run the FastAPI application with Uvicorn
+CMD ["uvicorn", "src.app:app", "--host", "0.0.0.0", "--port", "8000"]
```
Comment (pt20) [inline postgres/Dockerfile:24]: I'd rather put this command in the docker-compose file. (see: comment on compose file).

It provides more flexibility. You can easily change the command or its arguments without rebuilding the image. And you can reuse the same built image in different context (if needed)

File: postgres/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      PGADMIN_DEFAULT_EMAIL: ${PGADMIN_DEFAULT_EMAIL}
+      PGADMIN_DEFAULT_PASSWORD: ${PGADMIN_DEFAULT_PASSWORD}
+    volumes:
+      - ./pgadmin-data:/var/lib/pgadmin
+
+  fastapi-app:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    restart: always
```
Comment (pt20) [inline postgres/docker-compose.yaml:31]:
```suggestion
    restart: always
    command: uvicorn main:app --host 0.0.0.0 --port 8000
```

File: postgres/pgadmin/docker-compose.yaml
Comment (pt20) [inline postgres/pgadmin/docker-compose.yaml:1]: what's the purpose of this file?

File: postgres/Dockerfile
Comment (pt20) [inline postgres/Dockerfile:1]: You could also use dockerignore file. see https://shisho.dev/blog/posts/how-to-use-dockerignore/ for ref

File: postgres/src/config.py
Comment (pt20) [inline postgres/src/config.py:1]: nit: You could use pydantic's `BaseSettings` class to define your config. It has inbuilt functionalitzý to read from the env vars and a bunch of other improvement-of-life kind of functionality. Check this out. https://docs.pydantic.dev/1.10/usage/settings/

Comment (pt20) [reply postgres/src/config.py:1]: Note: this is pydantic 1.10. Recently they released pydantic v2 and I am not yet fully familiar with changes and differences.

Comment (pt20) [reply postgres/src/config.py:1]: Another advantage of using pydantic is that it comes with a postgresdsn type. So you can be always sure of the db url. https://docs.pydantic.dev/1.10/usage/types/#urls

File: postgres/src/app.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+from src.db import get_db
+from src.model import Task, Tasks
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+app = FastAPI()
+
+
+@app.post("/tasks", status_code=status.HTTP_201_CREATED)
```
Comment (pt20) [inline postgres/src/app.py:16]:
```suggestion
@app.post("/task", status_code=status.HTTP_201_CREATED)
```
At the moment, the app allows for sending only one task

Comment (pt20) [inline postgres/src/app.py:16]:
```suggestion
@app.post("/tasks", status_code=status.HTTP_201_CREATED, response_model=Task)
```

File: postgres/src/app.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        db.commit()
+        task_id = result.fetchone()[0]
+    except IntegrityError as e:
+        logger.warning(f"Error in creating task: {e.args[0]}")
+        if "psycopg2.errors.UniqueViolation" in e.args[0]:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Task already exists"
+            )
+        else:
+            raise HTTPException(
```
Comment (pt20) [inline postgres/src/app.py:30]: You can define a middleware to catch all the unanticipated errors. That way you won't need to clutter your code for every instance where some unexpected error may occur.

Comment (pt20) [reply postgres/src/app.py:30]: For inspiration - checkout the template that you should have.

File: postgres/src/db.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+url = postgres_url()
+
+# Create a SQLAlchemy engine
+engine = create_engine(url)
+
+# Create a session factory
+SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
+
+
+def get_db() -> Generator:
```
Comment (pt20) [inline postgres/src/db.py:17]: check fastapi `lifespan events` to do this more idiomatically. https://fastapi.tiangolo.com/advanced/events/

File: postgres/src/model.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from typing import List
+
+from pydantic import BaseModel
+
+
+class Task(BaseModel):
+    id: int
+    task: str
```
Comment (pt20) [inline postgres/src/model.py:8]: enhancement: You could also have a `timestamp`, `task_completion` and `status` of a task for bit more elaborate task.

File: postgres/src/model.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+
+class Tasks(BaseModel):
```
Comment (pt20) [inline postgres/src/model.py:11]: You don't necessarily neeed this. If you want to return list of tasks. You can define `response_model=List[Task]` and fastapi will do it's thing.

File: postgres/pgadmin/docker-compose.yaml
Comment (vikramsg) [reply postgres/pgadmin/docker-compose.yaml:1]: This was just a backup. I will remove.

File: postgres/src/config.py
Comment (vikramsg) [reply postgres/src/config.py:1]: Nice.

File: postgres/src/db.py (hunk 6 above)
Comment (vikramsg) [reply postgres/src/db.py:17]: I will have a look but the db session is more or less copied verbatim from FastAPI example. 
https://github.com/tiangolo/fastapi/blob/f7e3559bd5997f831fb9b02bef9c767a50facbc3/docs_src/sql_databases/sql_app/main.py#L15

**Pull request merged**